### ISSUE TITLE
Query enqueued assets (scripts and stylesheets) as connections

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#1308]
In WordPress, assets (scripts and stylesheets) can be enqueued and applied to the front end. 

In an application data graph, these styles and scripts can be considered nodes and they can have connections to other nodes, such as a Post, Page. 

As a user, I would like to be able to query for assets connected to existing resources. 

For example, given the following code to enqueue scripts in a WordPress plugin or theme: 

```
add_action( 'wp_enqueue_scripts', function() {

  // Register the scripts
  wp_register_script( 'equeue-everywhere', '/everywhere.js' );
  wp_register_script( 'only-single-posts', '/single-posts.js' );
  wp_register_script( 'only-single-categories', '/single-category.js' );
  wp_register_script( 'only-front-page', '/front-page.js' );
 
  // Enqueue the script EVERYWHERE
  wp_enqueue_script( 'equeue-everywhere' );

  // Enqueue just to single posts (of any post type)
  if ( is_single() ) {
     wp_enqueue_script( 'only-single-posts' );
  }

  // Enqueue to single categories
  if ( is_category() ) {
    wp_enqueue_script( 'only-single-categories' );
  }

  // Enqueue just on page set as the front page
  if ( is_front_page() ) {
    wp_enqueue_script( 'only-front-page' );
  }

} );
```

If I visited a single post on my front end, for example `site.com/post/hello-world`, I would see `/single-post.js` added as a `<script>` tag to that page. 

If I visited `site.com/` I would see `front-page.js` added as a `<script>` tag to the page, and the same with visiting `site.com/category/uncategorized` and seeing `/single-categories.js` added as a `<script>` tag. 

Often times, these scripts are used in conjunction with the content, to modify dom elements, such as convert unordered lists of images into slideshows, etc. 

In WPGraphQL, I would like to query these connected assets. 

Something like so: 

```graphql
query GET_POST_WITH_ENQUEUED_ASSETS {
  posts(first:1) {
    nodes {
       title
       enqueuedScripts {
         nodes {
           handle
           src
         }
       }
    }
  }
}
```

I might get a payload such as: 

```json
{
  "data": {
    "posts": {
        "nodes": [
            {
               "title": "hello-world",
               "enqueuedScripts": {
                   "nodes": [
                      {
                         "handle": "only-single-posts", 
                         "src": "/single-post.js"
                      },
                      {
                         "handle": "everywhere",
                         "src": "/enqueued-everywhere.js"
                      }
                    ]
                }
            }
        ]
    }
  }
}
```

This way, should I want to use these assets in my consuming application, I could ask for them and have them apply to my consuming application. 

I could do the same querying for terms, comments, etc. 

This will allow headless consumers to use the HTML output from the `post.content` field with the styles and interactions intended by the CMS (such as slideshows and other dynamic interactions).

**Comment by jasonbahl** [comment]
**NOTE:** have this partly working already: 

Using this PHP code in functions.php: 

```
add_action( 'wp_enqueue_scripts', function() {

  global $post;

  // Query to single posts in the "mouse" category
  if ( is_single( $post ) && in_category('mouse', $post->ID ) ) {
    wp_enqueue_script( 'only-single-post-mouse-category', '/single-post-mouse.js' );
  }

  // Query to all single posts
  if ( is_single() ) {
    wp_enqueue_script( 'only-single-post', '/single-post.js' );
  }

  // Enqueue to all pages
  if ( is_page() ) {
    wp_enqueue_script( 'only-page', '/single-page.js' );
  }

} );
```

I can query like so: 

```graphql
{
  contentNodes {
    nodes {
      __typename
      ... on NodeWithTitle {
        title
      }
      id
      enqueuedScripts {
        edges {
          node {
            id
            handle
            src
            version
            extra
          }
        }
      }
    }
  }
}
```

And I get this payload: 

```json
{
  "data": {
    "contentNodes": {
      "nodes": [
        {
          "__typename": "Post",
          "title": "Test post in mouse category",
          "id": "cG9zdDo1MzE=",
          "enqueuedScripts": {
            "edges": [
              {
                "node": {
                  "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLXBvc3QtbW91c2UtY2F0ZWdvcnk=",
                  "handle": "only-single-post-mouse-category",
                  "src": "/single-post-mouse.js",
                  "version": null,
                  "extra": null
                }
              },
              {
                "node": {
                  "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLXBvc3Q=",
                  "handle": "only-single-post",
                  "src": "/single-post.js",
                  "version": null,
                  "extra": null
                }
              }
            ]
          }
        },
        {
          "__typename": "Page",
          "title": "Test Page",
          "id": "cGFnZTo1Mjk=",
          "isSingular": false,
          "enqueuedScripts": {
            "edges": [
              {
                "node": {
                  "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktcGFnZQ==",
                  "handle": "only-page",
                  "src": "/single-page.js",
                  "version": null,
                  "extra": null
                }
              }
            ]
          }
        },
        {
          "__typename": "Post",
          "title": "Test Post NOT in mouse category",
          "id": "cG9zdDo1MDY=",
          "isSingular": true,
          "enqueuedScripts": {
            "edges": [
              {
                "node": {
                  "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLXBvc3Q=",
                  "handle": "only-single-post",
                  "src": "/single-post.js",
                  "version": null,
                  "extra": null
                }
              }
            ]
          }
        }
```

**Comment by TylerBarnes** [comment]
Awesome :D will there be a root query field to get a list of all enqueued assets?

**Comment by jasonbahl** [comment]
@TylerBarnes ideally. . .but it turns a lot of theme and plugin developers don't register scripts properly, so we'll have do at least document why this is an issue. 

A lot of folks do _just_ the following: 

```
add_action( 'wp_enqueue_scripts', function() {

  if ( is_single() ) {
    wp_enqueue_script( 'my-handle', '/path/to/my/script.js' );
  }

} );
```

This never actually adds the script to the global registry, so the script is _only_ made available when `is_single()` is true, and for a root query like: 

```
{
  enqueuedScripts {
     nodes {
       id
       handle
       src
     }
  }
}
```

The registry wouldn't actually have the script to return. 

So, we'll at least need to have documentation on this and encourage users to use `wp_register_script()` outside of any context like `is_single()`, etc and _only_ use the `wp_enqueue_script()` inside contextual arguments.

**Comment by TylerBarnes** [comment]
Makes sense!

**Comment by jasonbahl** [comment]
I think we should scope this initially to support connections to single nodes (posts/pages/cpt nodes, category/tag/custom tax nodes, comment nodes, menu nodes, menu item nodes)

Other contexts such as archive pages could theoretically be supported on the edge, like so, but initial implementation I think should probably focus on singular node->asset connections: 

```
{
  posts {
    enqueuedScripts { ...fields }
    nodes {
       id
       title
    }
  }
}
```

The following [conditions](https://codex.wordpress.org/Conditional_Tags) should be considered to test against as use cases users might use to enqueue scripts: 

- [ ] ~`is_home()`~
- [ ] `is_front_page()`
- [ ] `is_single()`
- [ ] `is_admin_bar_showing()` (potentially helpful for things like Gatsby Preview). . .not sure
- [ ] `is_single()`
- [ ] `is_sticky()` (maybe?)
- [ ] `is_post_type_hierarchical
- [ ] ~`is_post_type_archive()`~
- [ ] ~`is_comments_popup()`~
- [ ] `comments_open()`
- [ ] `pings_open()`
- [ ] `is_page()`
- [ ] `is_page_template()` 
- [ ] `is_category()` 
- [ ] `is_tag()` 
- [ ] `is_tax()`
- [ ] `has_term()`
- [ ] ~`term_exists()`~
- [ ] `is_taxonomy_hierarchical()`
- [ ] ~`taxonomy_exists()`
- [ ] `is_author()`
- [ ] ~`is_multi_author()`~
- [ ] ~`is_date()`~
- [ ] ~`is_year()`~
- [ ] ~`is_month()`~
- [ ] ~`is_day()`~
- [ ] ~`is_time()`~
- [ ] ~`is_new_day()`~
- [ ] ~`is_archive()`~
- [ ] ~`is_paged()`~ (we actually don't really support this at all in WPGraphQL right now. . .pagination on single pages. 🤔)
- [ ] `is_attachment()`
- [ ] `wp_attachment_is_image()` 
- [ ] `is_local_attachment()`
- [ ] `is_singular()`
- [ ] ~`post_type_exists()`~
- [ ] ~`is_main_query()`~
- [ ] ~`is_new_day()`~
- [ ] ~`is_feed()`~
- [ ] ~`is_trackback()`~
- [ ] `is_preview()`
- [ ] `has_excerpt()` 
- [ ] `has_nav_menu()`
- [ ] ~`in_the_loop()`~
- [ ] ~`is_dynamic_sidebar()`~
- [ ] ~`is_active_sidebar()`~
- [ ] ~`is_widget_active()`~
- [ ] ~`is_blog_installed()`~
- [ ] `is_rtl()`
- [ ] `is_multisite()`
- [ ] `is_main_site()` 
- [ ] `is_super_admin()`
- [ ] `is_user_logged_in()` 
- [ ] ~`email_exists()`~
- [ ] ~`username_exists()`~
- [ ] `is_plugin_active()`
- [ ] `is_child_theme()`
- [ ] `current_theme_supports()`
- [ ] `has_post_thumbnail()`
- [ ] `wp_script_is()`
- [ ] `is_customize_preview()`

**Comment by jasonbahl** [comment]
I've opened a PR with this working and tests for quite a few conditions. 

Here are some examples: 

### Enqeue Script
```php
add_action( 'wp_enqueue_scripts', function() {

	global $post;
 
        wp_register_script( 'only-single-post-mouse-category', '/single-post-mouse.js' );
        wp_register_script( 'only-single-any-post-type', '/single-any-post-type.js' );

        // Enqueue ONLY for Posts of the Post Type in the "mouse" category
	if ( is_singular( 'post' ) && in_category('mouse', $post->ID ) ) {
		wp_enqueue_script( 'only-single-post-mouse-category' );
	}

        // Enqueue to single posts of any post type
	if ( is_single() ) {
		wp_enqueue_script( 'only-single-any-post-type' );
	}
});
```

### Query for Posts and their enqueued scripts
```graphql
{
  posts {
    nodes {
      id
      title
      enqueuedScripts(first:2) {
        nodes {
          id
          handle
          src
        }
      }
    }
  }
}
```

### GraphQL Response
```json
{
  "data": {
    "posts": {
      "nodes": [
        {
          "id": "cG9zdDo1NDE=",
          "title": "Gatsby Debug",
          "enqueuedScripts": {
            "nodes": [
              {
                "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLWFueS1wb3N0LXR5cGU=",
                "handle": "only-single-any-post-type",
                "src": "/single-any-post-type.js"
              },
              {
                "id": "ZW5xdWV1ZWQtc2NyaXB0OnBpbmdzLW9wZW4=",
                "handle": "pings-open",
                "src": "/pings-open.js"
              }
            ]
          }
        },
        {
          "id": "cG9zdDo1MzE=",
          "title": "Mouse, yo",
          "enqueuedScripts": {
            "nodes": [
              {
                "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLXBvc3QtbW91c2UtY2F0ZWdvcnk=",
                "handle": "only-single-post-mouse-category",
                "src": "/single-post-mouse.js"
              },
              {
                "id": "ZW5xdWV1ZWQtc2NyaXB0Om9ubHktc2luZ2xlLWFueS1wb3N0LXR5cGU=",
                "handle": "only-single-any-post-type",
                "src": "/single-any-post-type.js"
              }
            ]
          }
        }
      ]
    }
  }
}
```

### Screenshot in GraphiQL
![Screen Shot 2020-05-13 at 2 16 16 PM](https://user-images.githubusercontent.com/1260765/81860988-c97d1300-9524-11ea-9148-7424f563dbde.png)

**Comment by jasonbahl** [comment]
Closed by #1313
